### PR TITLE
VirtualMachineReplicaSet to scale VMs

### PR DESCRIPTION
**Pull request by rmohr**
Initial simple implementation of a VirtualMachineReplicaSet.

Usage:

```yaml
apiVersion: kubevirt.io/v1alpha1
kind: VirtualMachineReplicaSet
metadata:
  name: myrs
spec:
  replicas: 3
  selector:
    matchLabels:
      myselector: myselector
  template:
    metadata:
      name: test
      labels:
        myselector: myselector
    spec:
      domain:
        devices:
          consoles:
          - type: pty
        memory:
          unit: MB
          value: 64
        os:
          type:
            os: hvm
        type: qemu
```

TODO:

- [x] proposal
- [x] logging
- [x] events
- [ ] graceful delete

### CODE REVIEW

File: docs/replica-sets.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+## Implementation
+
+A new object `VirtualMachineReplicaSet`, backed with a controller will be
+implemented:
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachineReplicaSet
+metadata:
+  name: replicasetn9z3w
```
Comment (stu-gott) [inline docs/replica-sets.md]: I'm confused, is this example the template or the manifest? If the former, this should be a generatedName, right?

File: pkg/virt-controller/watch/replicaset.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+				// TODO check if vm labels exist, and when make sure that they match. For now just override them
+				vm.ObjectMeta.Labels = rs.Spec.Template.ObjectMeta.Labels
+				_, err := c.clientset.VM(rs.ObjectMeta.Namespace).Create(vm)
+				if err != nil {
+					errChan <- err
+				}
+			}()
+		}
+	}
+	wg.Wait()
```
Comment (stu-gott) [inline pkg/virt-controller/watch/replicaset.go]: So this blocks until the changes to VM count have been scheduled?

File: pkg/kubecli/replicaset.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+func (v *rc) List(options k8smetav1.ListOptions) (replicasetList *v1.VirtualMachineReplicaSetList, err error) {
+	replicasetList = &v1.VirtualMachineReplicaSetList{}
+	err = v.restClient.Get().
+		Resource(v.resource).
+		Namespace(v.namespace).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Do().
+		Into(replicasetList)
+	for _, replicaset := range replicasetList.Items {
+		replicaset.SetGroupVersionKind(v1.VMReplicaSetGroupVersionKind)
```
Comment (stu-gott) [inline pkg/kubecli/replicaset.go]: gross :)

I assume you wouldn't have done this if there was a better way. Are we somehow requesting generic replicasets so they're not properly typed?

File: pkg/kubecli/replicaset.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+type rc struct {
+	restClient *rest.RESTClient
+	namespace  string
+	resource   string
+}
+
+func (v *rc) Get(name string, options k8smetav1.GetOptions) (replicaset *v1.VirtualMachineReplicaSet, err error) {
+	replicaset = &v1.VirtualMachineReplicaSet{}
+	err = v.restClient.Get().
```
Comment (stu-gott) [inline pkg/kubecli/replicaset.go]: Implicit assignment needs :=

Occurs a couple times here

File: pkg/kubecli/replicaset.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		Resource(v.resource).
+		Namespace(v.namespace).
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Do().
+		Into(replicaset)
+	replicaset.SetGroupVersionKind(v1.VMReplicaSetGroupVersionKind)
+	return
```
Comment (stu-gott) [inline pkg/kubecli/replicaset.go]: I'm assuming these returns that don't match function signature don't compile.

File: docs/replica-sets.md (hunk 1 above)
Comment (rmohr) [reply docs/replica-sets.md]: It is the name of the VirtualMachineReplicaSet. I can give it a nicer name.

File: pkg/kubecli/replicaset.go (hunk 5 above)
Comment (rmohr) [reply pkg/kubecli/replicaset.go]: had to adjust some functions to work with latest master, not that we updated client-go. Done.

File: pkg/kubecli/replicaset.go (hunk 3 above)
Comment (rmohr) [reply pkg/kubecli/replicaset.go]: That is actually a strange thing. The k8s clients don't fill that in. I think the proper solution would be to fill them in when we update or create a resource. Have to look into that again.

File: pkg/virt-controller/watch/replicaset.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: It could (I think the 8s ReplicaSet does), here I just do concurrent request on the apiserver, to not add up request delays, so that we can release the controller worker faster. I kind of borrowed that from the ReplicaSet controller from k8s.

File: pkg/kubecli/replicaset.go (hunk 4 above)
Comment (rmohr) [reply pkg/kubecli/replicaset.go]: Oh the function has named return types, No need to use ":=". The only reason why the code failed is because of the client-go update.

File: docs/replica-sets.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      name: test
+      labels:
+        mylabel: mylabel
+    spec:
+      domain:
+        devices:
+      [...]  
+```
+
+`spec.template` is similar to a `VirtualMachineSpec`. `spec.replicas` specifies
```
Comment (fabiand) [inline docs/replica-sets.md]: How `similar` is it? Wort specifying what it can/can not contain?

File: docs/replica-sets.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+It shows the number of `VirtualMachine`s which are in a non-final state nad
+which match `spec.selector`. In case of a scaling error, a `ReplicaFailure`
+condition is added to the status.
+
+In case of a delete failure:
+
+```yaml
+status:
+  conditions:
```
Comment (fabiand) [inline docs/replica-sets.md]: Conditions are great!

File: docs/replica-sets.md (hunk 6 above)
Comment (rmohr) [reply docs/replica-sets.md]: Actually not just similar, it is the same. Will fix the docu.

**Comment by rmohr** [comment]
@fabiand yes i considered reusing it. It is not yet fully from the table. So this can also be seen as a POC for the completely layered approach. Regarding to not having our own cluster level object, I think no matter if we reuse ReplicaSets, or not, we need that.

File: docs/replica-sets.md (hunk 6 above)
Comment (rmohr) [reply docs/replica-sets.md]: done

File: pkg/virt-controller/watch/replicaset.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+	var wg sync.WaitGroup
+	wg.Add(abs(diff))
+
+	if diff > 0 {
+		// We have to delete VMs
+		for i := 0; i < diff; i++ {
+			go func(idx int) {
+				defer wg.Done()
+				deleteCandidate := vms[idx]
```
Comment (davidvossel) [inline pkg/virt-controller/watch/replicaset.go]: In the future we may want to introduce delete strategies here, things like "delete oldest first"

File: pkg/virt-controller/watch/replicaset.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+	for _, obj := range objs {
+		rs := obj.(*virtv1.VirtualMachineReplicaSet)
+		replicaSets = append(replicaSets, *rs)
+	}
+	return replicaSets, nil
+}
+
+// vmChangeFunc checks if the supplied VM matches a replica set controller in it's namespace
+// and wakes the first controller which matches the VM labels.
+func (c *VMReplicaSet) vmChangeFunc(obj interface{}) {
```
Comment (davidvossel) [inline pkg/virt-controller/watch/replicaset.go]: One common pattern people use when replicating resources, whether it be containers or VMs, is the concept of isolating a resource from the group controller.

If someone manually changed the labels on a VM to isolate it from the replication controller, and then decides to terminate the resource, I don't believe the replication controller would ever get triggered to resize.

Basically, if someone changes a label on a VM and then deletes the VM, the VM replica set would be permanently at n-1 VMs until something else triggered the replica set to get processed.

**Comment by davidvossel** [comment]
@rmohr I agree that a having a VM specific cluster object to represent the replica set makes sense. I also like the idea of calling it a VirtualMachineReplicaSet. That naming convention makes it pretty obvious what the object does within the context of k8s.

I'm still contemplating the approach of implementing our own replication logic instead of leveraging the k8s controllers to do this for us. 

For the sake of argument, how would it work if we re-used the k8s replica set behind the scenes? Below are my thoughts on how I'd approach it..

- Creating a VirtualMachineReplicaSet (VMRS) would generate an internal k8s replica set
- The k8s replica set would start the VM pods.
- The VMRS controller would watch for pods related to itself and dynamically introduce VMs for each pod using the spec in the VMRS. These VMs should have a flag on them that prevents the VM controller from ever generating a POD for them.
- The VM controller would then see a VM got created, that corresponding POD already exists, and then proceed with assigning the VM to the node.
- When a POD related to a VMRS is completed/terminated/failed/whatever, the VMRS controller handles cleaning up the corresponding VM definition associated with it.

Basically, with this design the VMRS controller would just do these things.
- create/update/delete replica sets for corresponding VMRC objects.
- dynamically generate VM objectes as VMRS related pods appear
- remove VM objects as VMRC related pods disappear.

Thoughts on the Pros/Cons of something like what I've outlined above vs rolling out our own replication logic?

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
retest this please

File: pkg/virt-controller/watch/replicaset.go (hunk 8 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: Yep, that, or we can come up with a "best" sorting for that use-case. The k8s replica set does that on the pod level. Did not think about it in detail, though.

File: pkg/virt-controller/watch/replicaset.go (hunk 9 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: The ReplicaSet in k8s reacts to relabeling of Pods. If the labels do not match anymore, it either gets adopted by another matching ReplicaSet, or it stays in an orphaned state. The original ReplicaSet is woken up and will immediately create a new Pod.

Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: Actually, looks like I was wrong.

That [comment](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L272) indicates that it would wake up both controllers. Might be an artifact of the pre-controller-ref aera. However, the actual code in that function only reacts to label changes if the controller reference changes, or if the controller behind the reference was deleted. To detach a resource for sure from the controller (also prevent re-adoption by another controller), looks like one has to change labels and delete the controller ref. If only the labels are changed, it will still stick with the crating controller. If the controller reference is removed but the labels still stay the same it will be adopted again.

I will take that into account when adding the controller-ref. :+1:

Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: Ok, my original comment was right. The flow is as follows with controller-refs:

1) if labels changed, inform the owner only
2) the owner will try to release the object (remove the owner ref)
3) this causes another update on the resource. Now there is no owner, so all matching remaining controllers are informed
4) they race for adopting that resource

So changing a label on a Pod in the case of ReplicaSets leads to releasing the object and reclaiming it by another controlelr if one matches. Further the original ReplicaSet creates a new Pod  as a replacement for the old pod.

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
retest this please

**Comment by davidvossel** [comment]
I've seen the cloud-init tests timeout occasionally. I've increased the timeouts to match that of the console functional tests over here in this pr #444.

**Comment by rmohr** [comment]
retest this please

**Comment by fabiand** [comment]
Was there any reply to https://github.com/kubevirt/kubevirt/pull/453#issuecomment-330629711 ?

**Comment by rmohr** [comment]
So regarding to the approach: I think we are talking about potentially reusing three controllers ReplicaSet, StatefulSet and Deployment. The outlined steps above 

Regarding to pros/cons of reusing existing controllers:

pro:
 * In some cases it might mean a little less controller code
 * Initial code writing, like a VirtualMachineControllerRefManager can be delayed (however there already exist abstractions in client-go for most thinks)

con:
 * Loose control on what happens. E.g. a replicaset creates another pod already when a pod is marked for deletion. No matter if we want that, we just can't decide it.
 * Custom strategies for deletion orders, like mentioned by @davidvossel  are impossible to implement then
 * We would loose the ability to treat the VM as our atomic entity, if we change something internal to the VM, we have to go through each controller to see if it will work
 * If controllers have revisions, we still need to duplicate the revision logic
 * Listeners are still required, you would even have to listen to more entities. The base controller and pods, instead of just the VM.
 * Duplicating  the state to our controller from the k8s controller is still requried
 * Would we have to actually re-test every VM lifecycle feature to make sure that it works for each controller?
 * OwnerReferences still need to be maintained (pointing to the k8s core controller instead of the VMs).

At the end I have the feeling that it might be a little less code if we wrap around these three controllers, but to a high price. Since in both scenarios we need a lot of controller logic, I would actually prefer to keep a clear stacking at least for the ReplicaSet. If it turns out that it is an extreme amount of work to create something like a StatefulSet purely based on VMs, and therefore we create the infra to wrap around k8s core controllers, I would be open to discuss the VirtualMachineReplicaSet implementation again.

**Comment by davidvossel** [comment]
> If it turns out that it is an extreme amount of work to create something like a StatefulSet purely based on VMs, and therefore we create the infra to wrap around k8s core controllers, I would be open to discuss the VirtualMachineReplicaSet implementation again.

With this understanding, I'm in support @rmohr's initial approach to vm replica sets.

My biggest concern was that we were trivializing the amount of work to provide the kind of consistency the pod StatefulSets are aiming for. If we find that we're hitting some complex edge conditions that K8s already solves in their controllers, we should re-evaluate our implementation and see what we can reuse.

**Comment by rmohr** [comment]
State: Everything except graceful deletes and owner references is implemented. graceful deletes are generally not yet supported and tracked by #454. I want to add owner references in a follow up PR.

@davidvossel there is one surprise hidden in the new patches. I though that the expectation manager (https://github.com/kubevirt/kubevirt/pull/453/files#diff-aea407b8c3b5cc0af5f852ae77b4c54c) is already part of client-go. It was not. As a workaround I copied it over. It only depends on client-go and apimachinery dependencies. Will investigate if it can be exported to apimachinery.

File: pkg/virt-controller/watch/replicaset.go [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+			// The first matching rs will be returned
+			return &rs, nil
+		}
+
+	}
+	return nil, errors.NewNotFound(v1beta1.Resource("virtualmachinereplicaset"), "")
+}
+
+// addVirtualMachine searchs for a matching VMReplicaSet, updates it's expectations and wakes it up
+func (c *VMReplicaSet) addVirtualMachine(obj interface{}) {
```
Comment (davidvossel) [inline pkg/virt-controller/watch/replicaset.go]: add/del/update functions are all doing the same thing minus 1 line involving expectations.  Maybe create a helper function?

File: pkg/virt-controller/watch/replicaset.go [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+// getMatchingController returns the first VMReplicaSet which matches the labels of the VM from the listener cache.
+// If there are no matching controllers, a NotFound error is returned.
+func (c *VMReplicaSet) getMatchingController(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachineReplicaSet, error) {
+	log := logging.DefaultLogger()
+	controllers, err := c.listControllerFromNamespace(vm.ObjectMeta.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO check owner reference, if we have an existing controller which owns this one
```
Comment (davidvossel) [inline pkg/virt-controller/watch/replicaset.go]: can we do that now? or are we having to wait for something to get updated in client-go/apimachinery first?

File: pkg/virt-controller/watch/replicaset.go [hunk 12]
```diff
@@ -0,0 +1,533 @@
...
+					c.expectations.DeletionObserved(rsKey, kubecli.VirtualMachineKey(deleteCandidate))
+					c.recorder.Eventf(deleteCandidate, k8score.EventTypeWarning, FailedDeleteVirtualMachineReason, "Error deleting: %v", err)
+					errChan <- err
+					return
+				}
+				c.recorder.Eventf(deleteCandidate, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Deleted virtual machine: %v", deleteCandidate.ObjectMeta.UID)
+			}(i)
+		}
+
+	} else if diff < 0 {
```
Comment (davidvossel) [inline pkg/virt-controller/watch/replicaset.go]: Do we want to add any sort of protection for how many VMs we allow to be scaled in one batch?  I'm not sure where that kind of logic would belong. Is it something you've considered yet?

File: tests/replicaset_test.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+					}
+					break
+				}
+
+				Expect(err).ToNot(HaveOccurred())
+
+				Eventually(func() int32 {
+					rs, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Get(name, v12.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return rs.Status.Replicas
```
Comment (davidvossel) [inline tests/replicaset_test.go]: Right now, Status.Replicas may return the number of scheduled VMs, and not necessarily the number that have successfully started.

File: pkg/kubecli/expectations.go [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/* TODO:
+Taken from https://github.com/kubernetes/kubernetes/blob/1889a6ef52eb18b08e24843577c5b9d3b9a65daa/pkg/controller/controller_utils.go
```
Comment (davidvossel) [inline pkg/kubecli/expectations.go]: if we're going to bring in this code, we should bring in the unit tests that go with it as well.

File: tests/replicaset_test.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+	})
+
+	Context("A valid VirtualMachineReplicaSet given", func() {
+
+		table.DescribeTable("should scale", func(startScale int, stopScale int) {
+
+			template := tests.NewRandomVM()
```
Comment (davidvossel) [inline tests/replicaset_test.go]: should this vm have a read only disk?

File: docs/replica-sets.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    reason: "FailureCreate"
+    message: "no permission to create VMs"
+    lastTransmissionTime: "..."
+  replicas: 2
+```
+
+### Guarantees
+
+Once Graceful Deletes are implemented, the VirtualMachineReplicaSet guarantees
+that it will never create more than the requested numbers of VMs in a non-final
```
[the remark this answers is not in the document]
Comment (rmohr) [reply docs/replica-sets.md]: I think, that as soon as we have graceful delete (which means that only the administrator or virt-handler would do the final delete) we can give that guarantee. However, if no fencing action is taken and VMs start piling up in unknown or pending states, the replication controller will never create new VMs until the situation is resolved.

File: pkg/kubecli/expectations.go (hunk 14 above)
Comment (rmohr) [reply pkg/kubecli/expectations.go]: good point

File: tests/replicaset_test.go (hunk 13 above)
Comment (rmohr) [reply tests/replicaset_test.go]: Yes, not even scheduled ones. Just created replicas by the controller. It has a similar meaning like that field in the ReplicaSet. The ReplicaSet also contains fields available pods (ready state is true) and available (ready state for a specific amount of time). I can add them right away.

File: tests/replicaset_test.go (hunk 15 above)
Comment (rmohr) [reply tests/replicaset_test.go]: Yes. So far I never saw an issue because of that. We have many tests where we use the image like that. It is probably only a matter of time until we would see issues.

File: pkg/virt-controller/watch/replicaset.go (hunk 11 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: I would have to copy over the controller_ref_manager.go from k8s right now. It seems to be already perfectly well isolated there, but not yet in one of these libraries. 

We need a solution for all our controllers (vm, migration and this one) which follows the controller-ref proposal. It might be worth a separate PR. For as long as the user does not make labeling errors everything will work.

File: pkg/virt-controller/watch/replicaset.go (hunk 10 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: :+1:

File: pkg/virt-controller/watch/replicaset.go [hunk 17]
```diff
@@ -0,0 +1,533 @@
...
+		diff == 0 && scaleErr != nil && c.getCondition(rs, virtv1.VMReplicaSetReplicaFailure) != nil {
+		return nil
+	}
+
+	if scaleErr != nil {
+		// If an error occurred, only update to filtered pod count
+		rs.Status.Replicas = int32(len(vms))
+	} else {
+		// If no error occurred we have reached our required scale number
+		rs.Status.Replicas = int32(len(vms) - diff)
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: I think the fields the k8s ReplicaSet provides are pretty nice and sufficient. Mostly Replicas (with the meaning  it has in this controller), Ready and Active. I don't think we need pending, failed and succeeded here. If the garbage collector or a user would for instance delete pods in a final state, we can't provide an accurate number on re-synces. Pending would be an implicit value (Replicas - Ready).

File: pkg/virt-controller/watch/replicaset.go [hunk 18]
```diff
@@ -0,0 +1,533 @@
...
+		rs.Status.Conditions = []virtv1.VMReplicaSetCondition{}
+	}
+
+	_, err := c.clientset.ReplicaSet(rs.ObjectMeta.Namespace).Update(rs)
+	return err
+}
+
+func (c *VMReplicaSet) calcDiff(rs *virtv1.VirtualMachineReplicaSet, vms []virtv1.VirtualMachine) int {
+	// TODO default this on the aggregated api server
+	wantedReplicas := int32(1)
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: I wanted to give it the same default value like a ReplicaSet has, which is 1.

File: pkg/virt-controller/watch/replicaset.go (hunk 12 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: I didn't consider it, but extremely important! I think the protection can be achieved easily. With the expectations in place, we just need to limit ti diff to a maximum value in the for loops.

Then the controller will only create or delete pods again, after the added expectations are met.

File: pkg/virt-controller/watch/replicaset.go [hunk 19]
```diff
@@ -0,0 +1,533 @@
...
+				}
+				c.recorder.Eventf(deleteCandidate, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Deleted virtual machine: %v", deleteCandidate.ObjectMeta.UID)
+			}(i)
+		}
+
+	} else if diff < 0 {
+		// We have to create VMs
+		c.expectations.ExpectCreations(rsKey, abs(diff))
+		basename := c.getVirtualMachineBaseName(rs)
+		for i := diff; i < 0; i++ {
```
Comment (rmohr) [inline pkg/virt-controller/watch/replicaset.go]: Something like

```golang
for i := min(diff, burstReplicas); i < 0; i ++ {}
```

for creations and deletions should give us the protection.

File: pkg/virt-controller/watch/replicaset.go (hunk 12 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset.go]: Something like

```golang
for i := min(diff, burstReplicas); i < 0; i ++ {}
```

for creations and deletions should give us the protection. Also seems to be how the k8s controllers are limiting that. In the case of the ReplicaSet it is hardcoded to 500.

File: docs/replica-sets.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (smarterclayton) [reply docs/replica-sets.md]: Given that this is not a guarantee that ReplicaSets give it would be better to pick a different name.  A replica set is explictly created to mint fungible copies.  This guarantee is not part of what replicas do.

File: docs/replica-sets.md [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+state.
+
+As a consequence, in case of node failures, or in case of connection loss to
+virt-handlers which run VMs which are part of that set, no new VMs will be
+spawned until the VMs in unknown state are explicitly removed, or the node
+reconnects. This behaviour allows plugging in fencing controllers in a well
+defined way.
+
+It might make sense to weaken these guarantees in the future, and instead add a
+`StatefulSet` equivalent to KubeVirt in the future.
```
Comment (smarterclayton) [inline docs/replica-sets.md]: From an implementation perspective i was going to recommend that you *not* recreate the entire controller flow yourself, and instead simply spawn a new RS or SS with your desired state via the controller.  If you do that, you'll be limited to only the guarantees those controllers provide, but you won't have to reimplement the entire stack.

If you are planning on replicating the code stack, add some justifications here as to why.

Comment (davidvossel) [reply docs/replica-sets.md]: > From an implementation perspective i was going to recommend that you not recreate the entire controller flow yourself, and instead simply spawn a new RS or SS with your desired state via the controller.

@smarterclayton You might be right. This is something we've been discussing quite a bit this week.

Once we begin looking at what is involved with recreating the SS controller, I think there's a good chance we'll want to leverage the existing k8s controller over creating the controller flow ourselves. We don't have a clear understanding of the complexity involved or if our use case diverges from the k8s SS/RS controllers yet.

File: docs/replica-sets.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (rmohr) [reply docs/replica-sets.md]: Ok, good point. Will remove that guarantee.

**Comment by rmohr** [comment]
Everything except copying the unit test, and updating the proposal itself should be done now. Will update the proposal itself asap, including a justification for not just wrapping around the replication controller.

File: pkg/virt-controller/watch/replicaset.go (hunk 11 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/replicaset.go]: I see.  We'll want to follow up on this

File: pkg/virt-controller/watch/replicaset.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/virt-controller/watch/replicaset.go]: ah, okay. wfm then :+1:

File: pkg/virt-controller/watch/replicaset.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/virt-controller/watch/replicaset.go]: Okay, so "ReadyReplicas" are VMs in the running state, "Replicas" are vms that have be observed entering the system. 

It sounds reasonable to me to start with those two for now.

File: pkg/virt-controller/watch/replicaset_test.go [hunk 21]
```diff
@@ -0,0 +1,506 @@
...
+			// Count up on every invocation, so that we can measure how often this was called on one invocation
+			vmInterface.EXPECT().Create(gomock.Any()).Times(15).Do(func(arg interface{}) {
+				Expect(arg.(*v1.VirtualMachine).ObjectMeta.GenerateName).To(Equal("testvm"))
+			}).Return(vm, nil).Do(func(obj interface{}) {
+				invocations++
+			})
+
+			// Check if only 10 are created
+			controller.Execute()
+			Expect(invocations).To(Equal(10))
```
Comment (davidvossel) [inline pkg/virt-controller/watch/replicaset_test.go]: Are the mock callbacks somehow execute asynchronously?  This regression test failed during one of its runs, but passes at other times.

File: pkg/virt-controller/watch/replicaset.go (hunk 12 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/replicaset.go]: :+1:

File: docs/replica-sets.md [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+
+### Milestones
+
+ * Basic functionality
+ * Support graceful delete [1]
+ * Support controller references [2]
+ * Support label changes
+ * Support adopting orphaned Pods [2]
+ * Define a well known scale-down order
```
Comment (davidvossel) [inline docs/replica-sets.md]: Should 'Live Migration' of VMs in a replica set be a milestone?

**Comment by davidvossel** [comment]
Live migrations are an area where the VM replica set use case diverges from the container use case.

I don't think we have to have live migration completely solved within the context of replica sets before this work can be merged, but it would be great if we had a general idea of how we plan to approach that problem. @rmohr Maybe you could put your thoughts on this topic in the proposal?

File: pkg/virt-controller/watch/replicaset_test.go (hunk 21 above)
Comment (rmohr) [reply pkg/virt-controller/watch/replicaset_test.go]: Seems so. I am now just specifying the expected amount of invocations. That works.

**Comment by rmohr** [comment]
@smarterclayton added some justifications  why this implementation was chosen. Looking forward to hear your thoughts.

File: docs/replica-sets.md (hunk 22 above)
Comment (rmohr) [reply docs/replica-sets.md]: Just that it is mentioned that it should support migrations? Like the controller and migrations are implemented right now, live migration works with the controller.

**Comment by rmohr** [comment]
@davidvossel  added the explanation how I plan that we support live migrations for all types of controllers to the proposal. If we follow the currently layed out architecture (independent of the new discussion on the mailing list https://groups.google.com/forum/#!topic/kubevirt-dev/4XBlPPEcIq4), the live migration story is in my opinion solid an solved, or am I missing something?

File: docs/replica-sets.md [hunk 23]
```diff
@@ -98,6 +98,47 @@ unknown VirtualMachine states and graceful deletes, it might decide to already
 create new replicas in advance, to make sure that the amount of ready replicas
 stays close to the expected replica count.
 
+### Chosen Implementation
```
Comment (smarterclayton) [inline docs/replica-sets.md]: Naming wise, this does not behave like a ReplicaSet.  I would recommend that you spend some time looking for an alternate naming, because you provide guarantees that far exceed a Kubernetes ReplicaSet.  It will be likely confusing for users and admins: "A VMRS is like a ReplicaSet for pods except it's kind of like a StatefulSet except when it isn't".

File: docs/replica-sets.md [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+    reason: "FailureCreate"
+    message: "no permission to create VMs"
+    lastTransmissionTime: "..."
+  replicas: 2
+  readyReplicas: 3
+```
+
+### Guarantees
+
+The VirtualMachineReplicaSet  does **not** guarantee that there will never be
```
Comment (rmohr) [inline docs/replica-sets.md]: @smarterclayton Isn't this now pretty close to the guarantees a ReplicaSet gives? Maybe I just expressed it the wrong way. I wanted to weaken the guarantee to that of a ReplicaSet.

**Comment by rmohr** [comment]
@davidvossel added the tests for the expectations apart from the the discussion if we should keep the name for the controller, I think everything which should be here right now is there.

File: docs/replica-sets.md (hunk 24 above)
Comment (davidvossel) [reply docs/replica-sets.md]: The VMRS implementation behaves mostly like a RS now, but I see the potential for that to change. It seems possible we'll encounter use cases that require us to strengthen these guarantees in the future.  I'm doubtful we'll have to account for the same kind of strict ordering/persistence that SS aim to provide, but it's possible we'll have to diverge from the loose guarantees the k8s RS provides.

That said, at what point would our "enhanced" VMRS feature set confuse people that are already familiar with the k8s RS?  For me, as long as we plan to keep the VMRS focused on ephemeral VMs, I don't have a problem with drawing the comparison between VMRS and the k8s RS. The feature set between the two may diverge some, but conceptually they're both working with replicating ephemeral instances of immutable objects.

**Comment by davidvossel** [comment]
> @davidvossel added the tests for the expectations apart from the the discussion if we should keep the name for the controller, I think everything which should be here right now is there.

Really great work Roman!  This is solid.

Code and testing wise, everything looks in order to me at this point. I'm going to wait until we conclude the controller naming discussion before marking "approved"

**Comment by smarterclayton** [comment]
If you want to merge and continue the naming discussion that's fine - I
don't know that you have to do them in order as long as you think you can
resolve that concern before you get locked into supporting the API
